File: python3/120.py
class Solution:
    def minimumTotal(self, tri: List[List[int]]) -> int:
        def find(tri: List[List[int]], i: int, j: int, mins: List[int]) -> int:
            if i == len(tri) - 1:
                return tri[i][j]
            m1 = mins[i + 1] if (j != 0) else (find(tri, i + 1, j, mins))
            m2 = find(tri, i + 1, j + 1, mins)
            mins[i + 1] = m2
            return min(m1, m2) + tri[i][j]

        return find(tri, 0, 0, [0] * len(tri))

File: java/120.java
class Solution {
    public int minimumTotal(List<List<Integer>> tri) {
        int n = tri.size(), ps = 0;
        int mins[] = new int[n];
        return find(tri, 0, 0, mins);
    }

    private int find(List<List<Integer>> tri, int i, int j, int mins[]) {
        if (i == tri.size() - 1)
            return tri.get(i).get(j);
        int m1 = (j != 0) ? (mins[i + 1]) : (find(tri, i + 1, j, mins));
        int m2 = find(tri, i + 1, j + 1, mins);
        mins[i + 1] = m2;
        return Math.min(m1, m2) + tri.get(i).get(j);
    }
}
